### DISCUSSION TITLE
Docs: Change URL pattern to not break URLs with every patch release

### DISCUSSION THREAD

**Discussion by jpmckinney** [OP-TED/eForms-SDK#210] [Ideas]
The current situation is:

1. Version 1.3.0 is released. Docs are at https://docs.ted.europa.eu/eforms/1.3.0/index.html
2. Version 1.3.1 is released. Docs are at https://docs.ted.europa.eu/eforms/1.3.1/index.html. URLs with 1.3.0 now 404.

This is rather annoying as we constantly need to update our URLs (we regularly refer to eForms documentation in our own documentation).

There are two easy options:

1. Publish the 1.3 series at https://docs.ted.europa.eu/eforms/1.3/index.html In most patch releases (or maybe even all to date), there is no page that is added or removed, and no page that retains the same name yet fundamentally changes content. Even so, this option handles those scenarios gracefully. Switching to this option comes at no new cost, since, to date, you have already been breaking all the URLs with every patch release. This option is better than the above, because:
    1. URLs don't break.
    3. If a page is removed, the user gets a 404, just like above.
    4. If a page's content changes such that it is fundamentally different, it should be renamed. In this way, a user who linked to some page that was part of 1.3.0 won't load a page that is totally different when 1.3.1 is published. I'm sure you do this already (if it ever occurs), and so it is not any additional overhead for this proposal.
1. Keep the earlier 1.3.x documentation online. Honestly, I don't understand why it's removed from the internet entirely.

And there is one more difficult option:

3. Set up redirects. This requires a little more effort, but is also totally doable.

**Comment by rousso** [comment]
Hi @jpmckinney,

Try [https://docs.ted.europa.eu/eforms/**latest**](https://docs.ted.europa.eu/eforms/latest) :)
For example [https://docs.ted.europa.eu/eforms/**latest**/fields/index.html](https://docs.ted.europa.eu/eforms/latest/fields/index.html)

1. I will think about your idea. The problem with it, is that we may need to totally hide the patch number from everywhere in the site, which is something I am not very enthusiastic about. But I think there is a way around my problem that can also solve your problem too.

2. This is not a good solution for the specific issue.

3. The "/latest/" URL that I proposed above, is actually a redirect. I does not work on all pages yet but we will try to add it in every page we missed it. In general if we ever decide to move a page, we will use redirects. Currently we are in the process of migrating to Antora 3.x. When we are done with that we will revisit the "/latest/ redirect idea and try to make it universal.

Regards,
Yannis

**Reply by jpmckinney** [reply]
Can you add a redirect from /latest/schema/all-in-one.html ?

**Reply by rousso** [reply]
The redirect now works everywhere. See my other reply posted a minute ago at the bottom of this page.
The all-in-one.html is a pain and I intended to let it go. 
Do you really need the entire thing in one giant page?

**Reply by jpmckinney** [reply]
I find it very convenient to know that I can find anything on that page - and that if a schema element isn’t on that page, then it’s not anywhere else. The search bar is not always reliable in my experience. Before I found the all in one page, I would spend a lot of time visiting multiple results trying to find the most relevant page.

**Reply by rousso** [reply]
Well @jpmckinney , _"knowing that you can find anything on that page and that if a schema element isn't on that page, then it'snot anywhere else"_ is an assumption that you make which has no real basis. Nobody is maintaining that page. There are no guaranties that it will contain everything. It is only an intermediate step from our initial conversion of the docs from the original Word document where they were initially written, to the docs website. We just neglected to delete it until now. I still think it is better if we delete it. :)
The links do not really work in that page and it creates all kind of problems to us.

I have heard this argument before, (that having everything in one huge document somehow is more convenient to find stuff), but I am not very fond of the idea. This is a web based documentation. If people want just a PDF, then maybe we should have given everyone their PDF and stop worrying about all the complexities of maintaining a documentation portal. Why bother? Why do we even need web sites if PDFs are so convenient? :)

**Did you try the new docs search?** We changed it recently. It should be quite reliable now.
Let me know what you think about it.

**Reply by jpmckinney** [reply]
My understanding was that all-in-one was a concatenation of other pages. It sounds like that's not the case. That said, if the page *were* made to be generated from other content, then:

1. It would guarantee that it contained the same content as those other pages.
2. The links would work (unless there are repeated within-page anchor targets, I guess).

Anyway, on the search feature, my main interaction is looking up a BT or field. Let's say I search for a BT, like BT-777. The search results return each section and page it appears on. However, it doesn't provide any snippet or context, so I have no idea which result is mostly likely to contain the content I want. As such, I need to open multiple results.

When opening multiple results, I ctrl-click on a result to open it in a new window. The result listing then disappears AND the search box is cleared. I need to then re-enter my search terms to open the other results, while looking for the correct result. This is not really functional from a UX perspective...

Also, I get results across multiple versions of the SDK, and the versions are not in any comprehensible order. But I only ever want to search the most recent version. For BT-777 I need to scroll to the before-last group of results to see the 1.4.1 links (and I need to do this for each result I want to open). It'd be nice if, at least, the results were ordered from most to least recent version.

So, I agree that separate pages are OK, *on the condition* that the search interface is not a pain. Right now, it is a pain: in terms of lack of context, poor UX, and poor ordering.

**Reply by jpmckinney** [reply]
As such, I propose to keep the all-in-one page, as it seems like less effort than addressing the search issues.

**Reply by jpmckinney** [reply]
Oh, and I just tried to open the result on the elements.html page for "BT-777" in 1.4.1, but I get page not found...

**Reply by rousso** [reply]
Hi again,

The elements page is not needed. I removed it a couple of days ago. 
Like the all-in-one page, the elements page is just a duplication of content.

The only reason I did not yet remove the all-in-one page is because I am still thinking about what you said and trying to figure out what I can do to keep us both happy. :) My problem is that keeping the all-in-one page is against all design principles for the docs website. 

Did you try the search? We now use Algolia for search which is state of the art. What is your problem with the search? Is there something specific that bothers you with it, is or it that you generally prefer to click Ctrl-F and use find in page?

Let me know :)

**Reply by rousso** [reply]
> Oh, and I just tried to open the result on the elements.html page for "BT-777" in 1.4.1, but I get page not found...

I just checked and indeed the search index was not updated. 
I asked our devops engineer to correct the indexing script.

**Reply by jpmckinney** [reply]
> Did you try the search?

I described the significant UX issues above.

I know of Algolia; for example, https://getbootstrap.com/ uses it. However, in their implementation, actual content is highlighted in result snippets. For the eForms SDK, it's only headings that seem to get highlighted.

Bootstrap also only returns results from the version being viewed (good); they don't return results from all versions like eForms SDK does (not helpful).

That said, Bootstrap's implementation has the same issue of it being impossible to follow multiple results without re-entering the search terms, retrieving results, and clicking a new result. It would be better if the search modal stayed open after opening a result in a new tab. Edit: It seems if I right-click and open a new tab via the context menu, then the modal stays open. If I ctrl-click (or cmd-click) a result, then the modal closes.

**Reply by rousso** [reply]
Ok. Thanks for the feedback.
I will see what I can do with the ideas you provided.
I am keeping all-in-one for the moment. I removed it from the search results though so that we don't have duplicates there.
Thank you @jpmckinney

**Comment by jpmckinney** [comment]
Ah, great, I will try that. In my case, I will always be tracking latest. If any folks need to track earlier major/minor versions, option 1 might still be a good idea.

**Comment by rousso** [comment]
Hi again @jpmckinney,

Your suggestion has been implemented. Try it in th [docs website](https::docs.ted.europa.eu/eforms/latest).
So thank you for it!!

- Only major.minor version number is now in URLs.
- The "latest" redirect works on all pages.

Cheers,
Yannis

**Reply by jpmckinney** [reply]
Thank you! 🚀
